### PR TITLE
[SPARK-48664] Add Apache Spark 4.0.0-preview1 Dockerfiles

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

add docker files for 4.0.0-preview1

### Why are the changes needed?

new release

### Does this PR introduce _any_ user-facing change?

new release

### How was this patch tested?

new tests

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @Yikun @yaooqinn

**Comment by yaooqinn** [comment]
Thank you @cloud-fan for publishing the docker image, can we make the CI pass?

**Comment by cloud-fan** [comment]
```
27.02 gpg: Can't check signature: No public key
```
I have no idea why it fails...

**Comment by cloud-fan** [comment]
I just followed https://github.com/apache/spark-docker/pull/33

File: 4.0.0-preview1/scala2.13-java17-ubuntu/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+ENV SPARK_TGZ_URL=https://archive.apache.org/dist/spark/spark-4.0.0-preview1/spark-4.0.0-preview1-bin-hadoop3.tgz \
+    SPARK_TGZ_ASC_URL=https://archive.apache.org/dist/spark/spark-4.0.0-preview1/spark-4.0.0-preview1-bin-hadoop3.tgz.asc \
+    GPG_KEY=4DC9676CEF9A83E98FCA02784D6620843CD87F5A
+
+RUN set -ex; \
+    export SPARK_TMP="$(mktemp -d)"; \
+    cd $SPARK_TMP; \
+    wget -nv -O spark.tgz "$SPARK_TGZ_URL"; \
+    wget -nv -O spark.tgz.asc "$SPARK_TGZ_ASC_URL"; \
+    export GNUPGHOME="$(mktemp -d)"; \
```
Comment (yaooqinn) [inline 4.0.0-preview1/scala2.13-java17-ubuntu/Dockerfile:48]:
```suggestion
    export GNUPGHOME="$(mktemp -d)"; \
    wget -nv -O KEYS https://downloads.apache.org/spark/KEYS; \
    gpg --import KEYS; \
```

File: 4.0.0-preview1/scala2.13-java17-ubuntu/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    wget -nv -O KEYS https://downloads.apache.org/spark/KEYS; \
```
Comment (cloud-fan) [inline 4.0.0-preview1/scala2.13-java17-ubuntu/Dockerfile:49]: @yaooqinn do you know why other Spark versions' dockerfiles are fine without this change?

Comment (yaooqinn) [reply 4.0.0-preview1/scala2.13-java17-ubuntu/Dockerfile:49]: I'm not sure, (it might because your key is new and theirs are cached)

**Comment by cloud-fan** [comment]
```
docker: Error response from daemon: failed to create task for container: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: exec: "/opt/entrypoint.sh": permission denied: unknown.
```
let me check the file permissions.

**Comment by yaooqinn** [comment]
It seems that we need to bump up mini kube to fix ITs like https://github.com/apache/spark/pull/44813

File: .github/workflows/main.yml [hunk 3]
```diff
@@ -262,7 +262,7 @@ jobs:
           # See more in "Installation" https://minikube.sigs.k8s.io/docs/start/
           # curl -LO https://storage.googleapis.com/minikube/releases/latest/minikube-linux-amd64
           # TODO(SPARK-44495): Resume to use the latest minikube for k8s-integration-tests.
```
Comment (yaooqinn) [inline .github/workflows/main.yml:264]: We can remove these comments

**Comment by cloud-fan** [comment]
The merge script didn't work for me, I'll just use the merge PR button in github.

**Comment by cloud-fan** [comment]
@Yikun I think we need to update `add-dockerfiles.sh`. Spark 4.x uses scala 2.13 now.

File: versions.json [hunk 4]
```diff
@@ -1,5 +1,34 @@
 {
   "versions": [
+    {
+      "path": "4.0.0-preview1/scala2.13-java17-python3-ubuntu",
+      "tags": [
+        "4.0.0-preview1-scala2.13-java17-python3-ubuntu",
+        "4.0.0-preview1-java17-python3",
+        "4.0.0-preview1-java17",
+        "python3-java17"
```
Comment (cloud-fan) [inline versions.json:9]: With hindsight, I think we should have a more strict naming rule for tags. For example, `python3-java17` is a bad name as we don't even know which spark version it is.

I'd propose: `spark_version-scala_version-java_version-[python support]-[r support]-OS`

if the spark version only supports one scala or java version, we can omit the scala/java version in the name.

What do you think? @Yikun @yaooqinn

Comment (yaooqinn) [reply versions.json:9]: +1 for the name. 

Is -OS necessary as a JVM app or not？If necessary, the OS versions or codenames like `Focal Fossa` also matter?

Comment (cloud-fan) [reply versions.json:9]: ah we can omit OS name as well. If one day we publish docker images with different OS, we can add back the OS in the tag name.

**Comment by yaooqinn** [comment]
I have seen these tags present at https://hub.docker.com/r/apache/spark/tags but absent at https://hub.docker.com/_/spark

It seems that we need to a PR like https://github.com/docker-library/official-images/pull/16325 to publish Docker Official Images

**Comment by yaooqinn** [comment]
https://hub.docker.com/layers/apache/spark/latest/images/sha256-4494ef2d6cb94e343402d67a0151ab39ce9b1a13810b82e5cbe55530c895aaf7?context=explore

The version which the latest tag points to is also outdated

**Review by Yikun — COMMENTED** [review]
So sorry for the late reply

File: versions.json (hunk 4 above)
Comment (cloud-fan) [reply versions.json:9]: It's hard to believe people don't care about spark versions...

Comment (Yikun) [reply versions.json:9]: So, the idea is  should we want to remove `python3`, `scala`, `r` special tag?

original:
```
latest: the latest version of current majar.minor version
3.3.0: the 3.3.0 version with default scala/python3
python3/r/scala: latest python3/r/scala version
3.3.0-r/3.3.0-scala/3.3.0-python3:  the 3.3.0 version with r/scala only/python3
```

tobe:
```
latest: the latest version of current majar.minor version
3.3.0: the 3.3.0 version with default scala/python3
3.3.0-r/3.3.0-scala/3.3.0-python3:  the 3.3.0 version with r/scala only/python3
```

> It's hard to believe people don't care about spark versions...

- IIRC, the background of `python3/r/scala` is that there were 3 images before (spark, spark-py, spark-r), then spark --> scala, spark-py --> python3, spark-r --> r .
- BTW, the `spark:python3`, `spark:r` is also used by https://spark.apache.org/

also cc @HyukjinKwon

Comment (cloud-fan) [reply versions.json:9]: I'm wondering if we should provide a single image with all supports (Python and R). I don't think package size is a big concern here. We can also provide variants with different scala and java versions.

Comment (cloud-fan) [reply versions.json:9]: yea, the all-in-one image isn't that big. Why do we provide other variants that remove python or r support?

Comment (Yikun) [reply versions.json:9]: There are 2 reason I think:
1. To cover [kubernetes best practice](https://cloud.google.com/blog/products/containers-kubernetes/kubernetes-best-practices-how-and-why-to-build-small-container-images), to keep images size as small as possible.
2. As I said before, there were spark-py / spark-r / spark before, to align old version images.

**Pull request merged**